spec/polymorphic_spec.cr

```diff
@@ -0,0 +1,123 @@
+require "./spec_helper"
+
+include LazyLoadHelpers
+
+private class PolymorphicTask < BaseModel
+  table do
+    column title : String
+  end
+end
+
+private class PolymorphicTaskList < BaseModel
+  table do
+    column title : String
+  end
+end
+
+private class PolymorphicEvent < BaseModel
+  table do
+    belongs_to task : PolymorphicTask?
+    belongs_to task_list : PolymorphicTaskList?
+    polymorphic :eventable, associations: [:task, :task_list]
+  end
+end
+
+private class OptionalPolymorphicEvent < BaseModel
+  table :polymorphic_events do
+    belongs_to task : PolymorphicTask?
+    belongs_to task_list : PolymorphicTaskList?
+    polymorphic :optional_eventable, optional: true, associations: [:task, :task_list]
+  end
+end
+
+describe "polymorphic belongs to" do
+  it "sets up a method for accessing associated record" do
+    task = PolymorphicTask::SaveOperation.create!(title: "Use Lucky")
+    event = PolymorphicEvent::SaveOperation.create!(task_id: task.id)
+    event.eventable.should eq(task)
+
+    task_list = PolymorphicTaskList::SaveOperation.create!(title: "Use Lucky")
+    event = PolymorphicEvent::SaveOperation.create!(task_list_id: task_list.id)
+    event.eventable.should eq(task_list)
+  end
+
+  it "can require preloading" do
+    with_lazy_load(enabled: false) do
+      expect_raises Avram::LazyLoadError do
+        task = PolymorphicTask::SaveOperation.create!(title: "Use Lucky")
+        event = PolymorphicEvent::SaveOperation.create!(task_id: task.id)
+        event.eventable # should raise because it was not preloaded
+      end
+    end
+  end
+
+  it "has ! method to allow lazy loading" do
+    with_lazy_load(enabled: false) do
+      task = PolymorphicTask::SaveOperation.create!(title: "Use Lucky")
+      event = PolymorphicEvent::SaveOperation.create!(task_id: task.id)
+      event.eventable!.should eq(task)
+    end
+  end
+
+  it "can preload the polymorphic associations" do
+    with_lazy_load(enabled: false) do
+      task = PolymorphicTask::SaveOperation.create!(title: "Use Lucky")
+      event = PolymorphicEvent::SaveOperation.create!(task_id: task.id)
+      event = PolymorphicEvent::BaseQuery.new.preload_eventable.find(event.id)
+      event.eventable.should eq(task)
+
+      # Check that it preloads both belongs_to, not just the first
+      task_list = PolymorphicTaskList::SaveOperation.create!(title: "Use Lucky")
+      event = PolymorphicEvent::SaveOperation.create!(task_list_id: task_list.id)
+      event = PolymorphicEvent::BaseQuery.new.preload_eventable.find(event.id)
+      event.eventable.should eq(task_list)
+    end
+  end
+
+  describe "required (default)" do
+    it "validates that exactly one polymorphic belongs to is allowed" do
+      operation = PolymorphicEvent::SaveOperation.create(task_id: 1, task_list_id: 1) do |operation, _event|
+        operation.valid?.should be_false
+        operation.task_list_id.errors.should eq(["must be blank"])
```

Comment:
Sounds tricky to do, but what if the error message was on the operation instead:

> When using polymorphic associations, only one model may be associated at once.
>
> Models associated currently:
> - Task
> - TaskList

Reply:
Yeah the more I think about this the harder it seems. I think I'm just looking for a hint of where to look if I get this error. What about a field error:

```crystal
errors.should eq(["can't be set if `task_id` is also set"])
```

Reply:
I like this idea a lot but these error messages might be shown to users if you serialize all the errors in a JSON response for example. I think what we need is something like #15 so we can add  more helpful errors like this and display them in a nice way to developers